src/librustc/middle/const_eval.rs

```diff
@@ -361,10 +538,13 @@ pub fn eval_const_expr_partial<'tcx>(tcx: &ty::ctxt<'tcx>,
               ast::BiBitOr => a | b,
               ast::BiEq => a == b,
               ast::BiNe => a != b,
-              _ => return Err("can't do this op on bools".to_string())
+              _ => return Err(invalid_op_for_bools(e, op.node)),
              }))
           }
-          _ => Err("bad operands for binary".to_string())
+          (err @ Err(..), _) |
+          (_, err @ Err(..)) => err,
+
+          _ => Err(misc_binary_op(e)),
```

Comment:
My bad, I was reading the pattern above wrong, I thought it was only matching `(Ok(_), Err(_)) | (Err(_), Ok(_))`.
I guess it's equivalent to using `try!`, so why not do that and get rid of some `Ok` pattern noise?